File: app/src/main/java/com/pr0gramm/app/ui/views/viewer/ProxyMediaView.java
package com.pr0gramm.app.ui.views.viewer;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.support.annotation.NonNull;
import android.view.MotionEvent;
import android.view.View;

import com.pr0gramm.app.R;

/**
 */
@SuppressLint("ViewConstructor")
public abstract class ProxyMediaView extends MediaView {
    private MediaView child;

    public ProxyMediaView(Activity context, MediaUri uri, Runnable onViewListener) {
        super(context, R.layout.player_proxy, uri, onViewListener);
        showBusyIndicator();
    }

    public void setChild(MediaView child) {
        removeChildView();
        hideBusyIndicator();

        setChildView(child);

        bootupChild();

        // forward double clicks
        child.setTapListener(new ForwardingTapListener());
    }

    /**
     * Adds the proxied child above the preview.
     */
    private void setChildView(MediaView mediaView) {
        int idx = getChildCount();
        View previewView = getPreviewView();
        if (previewView != null && previewView.getParent() == this) {
            idx = indexOfChild(previewView) + 1;
        }

        // transfer the layout parameters
        mediaView.setLayoutParams(getLayoutParams());
        mediaView.setViewAspect(getViewAspect());
        addView(mediaView, idx);

        child = mediaView;
    }

    private void removeChildView() {
        if (child == null)
            return;

        teardownChild();
        removeView(child);

        child = null;
    }

    private void bootupChild() {
        if (child != null) {
            if (isResumed())
                child.onResume();

            if (isPlaying())
                child.playMedia();
        }
    }

    private void teardownChild() {
        if (child != null) {
            if (isPlaying())
                child.stopMedia();

            if (isResumed())
                child.onPause();
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        if (child != null)
            child.onResume();
    }

    @Override
    public void playMedia() {
        super.playMedia();
        if (child != null)
            child.playMedia();
    }

    @Override
    public void stopMedia() {
        super.stopMedia();
        if (child != null)
            child.stopMedia();
    }

    @Override
    public void rewind() {
        if (child != null)
            child.rewind();
    }

    @Override
    public MediaView getActualMediaView() {
        return child != null ? child : this;
    }

    @Override
    protected void onMediaShown() {
        setViewAspect(-1);
        removePreviewImage();
        super.onMediaShown();
    }

    @Override
    public void onPause() {
        super.onPause();
        if (child != null)
            child.onPause();
    }

    @Override
    public boolean onTouchEvent(@NonNull MotionEvent event) {
        if (child != null) {
            event.offsetLocation(
                    child.getPaddingLeft() - getPaddingLeft(),
                    child.getPaddingTop() - getPaddingTop());

            return child.onTouchEvent(event);
        }

        return super.onTouchEvent(event);
    }

    private class ForwardingTapListener implements TapListener {
        @Override
        public boolean onSingleTap() {
            return getTapListener() != null && getTapListener().onSingleTap();
        }

        @Override
        public boolean onDoubleTap() {
            return getTapListener() != null && getTapListener().onDoubleTap();
        }
    }
}

File: app/src/main/java/com/pr0gramm/app/ui/views/viewer/ProxyMediaView.kt
package com.pr0gramm.app.ui.views.viewer

import android.annotation.SuppressLint
import android.view.MotionEvent
import com.pr0gramm.app.R
import com.pr0gramm.app.ui.base.whileIsAttachedScope
import kotlinx.coroutines.launch

/**
 */
@SuppressLint("ViewConstructor")
abstract class ProxyMediaView internal constructor(config: Config) : MediaView(config, R.layout.player_proxy) {
    private var delegate: MediaView? = null

    init {
        showBusyIndicator()
    }

    internal fun setChild(child: MediaView) {
        hideBusyIndicator()

        var idx = childCount

        val previewView = previewView
        if (previewView?.parent === this) {
            idx = indexOfChild(previewView) + 1
        }

        child.layoutParams = layoutParams
        child.viewAspect = viewAspect

        addView(child, idx)

        delegate = child

        if (isPlaying) {
            child.playMedia()
        }

        // forward double clicks
        child.tapListener = ForwardingTapListener()

        child.wasViewed = { onMediaShown() }

        whileIsAttachedScope {
            launch {
                child.controllerViews().collect { view -> publishControllerView(view) }
            }
        }
    }

    override fun playMedia() {
        super.playMedia()
        delegate?.playMedia()
    }

    override fun stopMedia() {
        super.stopMedia()
        delegate?.stopMedia()
    }

    override fun rewind() {
        delegate?.rewind()
    }

    override val actualMediaView: MediaView
        get() = delegate ?: this

    override fun onMediaShown() {
        viewAspect = -1f
        removePreviewImage()
        super.onMediaShown()
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        delegate?.let { child ->
            event.offsetLocation(
                    (child.paddingLeft - paddingLeft).toFloat(),
                    (child.paddingTop - paddingTop).toFloat())

            return child.onTouchEvent(event)
        }

        return super.onTouchEvent(event)
    }

    private inner class ForwardingTapListener : TapListener {
        override fun onSingleTap(event: MotionEvent): Boolean {
            return tapListener?.onSingleTap(event) ?: false
        }

        override fun onDoubleTap(event: MotionEvent): Boolean {
            return tapListener?.onDoubleTap(event) ?: false
        }
    }
}
